### PR TITLE
Update Init_orca_context

### PR DESCRIPTION
**Pull request by sgwhat**

### CODE REVIEW

**Comment by hkvision** [comment]
@jason-dai Shall we directly call init_nncontext in init_orca_context? As dllib has all these things now.

File: python/orca/src/bigdl/orca/common.py [hunk 1]
```diff
@@ -158,21 +158,18 @@ def get_ray_context():
...
     across the cluster if necessary).
-
     :param cluster_mode: The mode for the Spark cluster. One of "local", "yarn-client",
            "yarn-cluster", "k8s-client", "standalone" and "spark-submit". Default to be "local".
-
            For "spark-submit", you are supposed to use spark-submit to submit the application.
            In this case, please set the Spark configurations through command line options or
            the properties file. You need to use "spark-submit" for yarn-cluster or k8s-cluster mode.
            To make things easier, you are recommended to use the launch scripts we provide:
            https://github.com/intel-analytics/analytics-zoo/tree/master/scripts.
```
Comment (hkvision) [inline python/orca/src/bigdl/orca/common.py:172]: Also need to modify these? at least remove analytics-zoo...

File: python/orca/src/bigdl/orca/common.py [hunk 2]
```diff
@@ -183,79 +180,95 @@ def init_orca_context(cluster_mode="local", cores=2, memory="2g", num_nodes=1,
            Default to be False and in this case the Ray cluster would be launched lazily when
            Ray is involved in Project Orca.
     :param kwargs: The extra keyword arguments used for creating SparkContext and
-           launching Ray if any.
-
+           launching Ray if any. 
+           For k8s-client mode, you should specify master as ${SPARK_MASTER} (k8s://https://localhost:8443)
+           and set container_image to ${K8s_SPARK_IMAGE}.
```
Comment (hkvision) [inline python/orca/src/bigdl/orca/common.py:185]: Put this in cluster mode?
```
For k8s-client mode, you should additionally specify the arguments master and container_image.
```

Comment (sgwhat) [reply python/orca/src/bigdl/orca/common.py:185]: Better to keep this in client mode, I didn't see any arg is required in cluster mode(spark-submit).

File: python/orca/src/bigdl/orca/common.py (hunk 1 above)
Comment (sgwhat) [reply python/orca/src/bigdl/orca/common.py:172]: Sure we need.

File: python/orca/src/bigdl/orca/common.py (hunk 2 above)
Comment (hkvision) [reply python/orca/src/bigdl/orca/common.py:185]: I mean cluster_mode argument....
A similar paragraph like `For "spark-submit", you are supposed to use spark-submit to submit the application.....` here https://github.com/intel-analytics/BigDL/pull/3397/files#diff-e52466efd112054f4c0e38a19903bc5e8e40a3b9caa3291c7b67110069b2d912R168

Comment (sgwhat) [reply python/orca/src/bigdl/orca/common.py:185]: > I mean cluster_mode argument.... A similar paragraph like `For "spark-submit", you are supposed to use spark-submit to submit the application.....` here https://github.com/intel-analytics/BigDL/pull/3397/files#diff-e52466efd112054f4c0e38a19903bc5e8e40a3b9caa3291c7b67110069b2d912R168

Haha sorry, that's a good idea.

File: python/orca/src/bigdl/orca/common.py [hunk 3]
```diff
@@ -183,79 +180,95 @@ def init_orca_context(cluster_mode="local", cores=2, memory="2g", num_nodes=1,
...
             spark_args[key] = kwargs[key]
-    if cluster_mode == "spark-submit":
+    if cluster_mode is not None:
+        cluster_mode = cluster_mode.lower()
+    activate_sc = SparkContext._active_spark_context is not None
+    if activate_sc:
+        if cluster_mode is not None:
+            warnings.warn("Use an existing SparkContext, " +
+                          "cluster_mode is determined by the existing SparkContext", Warning)
+        conf = SparkConf()
```
[the remark this answers is not in the document]
Comment (sgwhat) [reply python/orca/src/bigdl/orca/common.py:204]: > is it necessary? As you dig into the code, https://github.com/intel-analytics/BigDL/blob/branch-2.0/python/dllib/src/bigdl/dllib/nncontext.py#L678 it will create a SparkConf by default, yes?

Yes, it could create a sparkconf object same as system default, which is necessary in https://github.com/sgwhat/BigDL/blob/755b04f3ce023aab0adae5c9292bbe7f64ad7f81/python/orca/src/bigdl/orca/common.py#L206.

**Comment by jason-dai** [comment]
> @jason-dai Shall we directly call init_nncontext in init_orca_context? As dllib has all these things now. It is better if we directly call init_nncontext and modify in init_nncontext to avoid duplicate work? cc @dding3 @qiuxin2012

Yes - but probably in a separate PR

File: python/orca/src/bigdl/orca/common.py [hunk 4]
```diff
@@ -183,79 +182,93 @@ def init_orca_context(cluster_mode="local", cores=2, memory="2g", num_nodes=1,
...
-                                   executor_memory=memory, **spark_args)
+        sc = init_nncontext(conf=conf, spark_log_level="WARN", redirect_spark_log=True)
     else:
-        raise ValueError("cluster_mode can only be local, yarn-client, standalone or spark-submit, "
-                         "but got: %s".format(cluster_mode))
+        cluster_mode = "local" if cluster_mode is None else cluster_mode
+        if cluster_mode == "spark-submit":
+            raise ValueError("spark-submit has been deprecated")
+        elif cluster_mode == "local":
+            assert num_nodes == 1, "For Spark local mode, num_nodes should be 1"
```
Comment (qiuxin2012) [inline python/orca/src/bigdl/orca/common.py:212]: just give a warning when the num_nodes > 1

File: python/orca/src/bigdl/orca/common.py [hunk 5]
```diff
@@ -183,79 +182,93 @@ def init_orca_context(cluster_mode="local", cores=2, memory="2g", num_nodes=1,
...
+            os.environ["SPARK_DRIVER_MEMORY"] = memory
+            if "python_location" in kwargs:
+                spark_args["python_location"] = kwargs["python_location"]
+            from bigdl.dllib.nncontext import init_spark_on_local
+            sc = init_spark_on_local(cores, **spark_args) 
+        elif cluster_mode.startswith("yarn"):  # yarn or yarn-client
+            if cluster_mode == "yarn-cluster":
+                raise ValueError('For yarn-cluster mode, '
```
Comment (qiuxin2012) [inline python/orca/src/bigdl/orca/common.py:220]: `yarn-cluster` is supported by #3313

File: python/orca/src/bigdl/orca/common.py [hunk 6]
```diff
@@ -158,22 +158,21 @@ def get_ray_context():
...
     across the cluster if necessary).
-
     :param cluster_mode: The mode for the Spark cluster. One of "local", "yarn-client",
            "yarn-cluster", "k8s-client", "standalone" and "spark-submit". Default to be "local".
-
            For "spark-submit", you are supposed to use spark-submit to submit the application.
            In this case, please set the Spark configurations through command line options or
            the properties file. You need to use "spark-submit" for yarn-cluster or k8s-cluster mode.
+           For "k8s-client", you are supposed to additionally specify the arguments master 
+           and container_image.
```
Comment (hkvision) [inline python/orca/src/bigdl/orca/common.py:172]: Put this before For spark-submit and add a blank line, since the next paragraph goes with spark-submit

File: python/orca/src/bigdl/orca/common.py [hunk 7]
```diff
@@ -183,79 +182,93 @@ def init_orca_context(cluster_mode="local", cores=2, memory="2g", num_nodes=1,
            Default to be False and in this case the Ray cluster would be launched lazily when
            Ray is involved in Project Orca.
     :param kwargs: The extra keyword arguments used for creating SparkContext and
-           launching Ray if any.
-
+           launching Ray if any.
```
Comment (hkvision) [inline python/orca/src/bigdl/orca/common.py:185]: revert this change

File: python/orca/src/bigdl/orca/common.py (hunk 5 above)
Comment (hkvision) [reply python/orca/src/bigdl/orca/common.py:220]: will update in a following PR to combine with init nncontext.

File: python/orca/src/bigdl/orca/common.py [hunk 8]
```diff
@@ -158,22 +158,25 @@ def get_ray_context():
...
 
-def init_orca_context(cluster_mode="local", cores=2, memory="2g", num_nodes=1,
+def init_orca_context(cluster_mode=None, cores=2, memory="2g", num_nodes=1,
                       init_ray_on_spark=False, **kwargs):
     """
     Creates or gets a SparkContext for different Spark cluster modes (and launch Ray services
     across the cluster if necessary).
 
     :param cluster_mode: The mode for the Spark cluster. One of "local", "yarn-client",
            "yarn-cluster", "k8s-client", "standalone" and "spark-submit". Default to be "local".
```
Comment (hkvision) [inline python/orca/src/bigdl/orca/common.py:168]: Need to change the doc here. Default to be None?

File: python/orca/src/bigdl/orca/common.py [hunk 9]
```diff
@@ -158,22 +158,25 @@ def get_ray_context():
...
+           
+           For "k8s-client", you are supposed to additionally specify the arguments master 
+           and container_image.
 
            For "spark-submit", you are supposed to use spark-submit to submit the application.
```
Comment (hkvision) [inline python/orca/src/bigdl/orca/common.py:173]: Change here?

File: python/orca/src/bigdl/orca/common.py [hunk 10]
```diff
@@ -183,79 +186,92 @@ def init_orca_context(cluster_mode="local", cores=2, memory="2g", num_nodes=1,
...
+            for key in ["driver_cores", "driver_memory", "extra_executor_memory_for_ray",
+                        "extra_python_lib", "jars", "master", "python_location", "enable_numa_binding"]:
+                if key in kwargs:
+                    spark_args[key] = kwargs[key]
+            from bigdl.dllib.nncontext import init_spark_standalone
+            sc = init_spark_standalone(num_executors=num_nodes, executor_cores=cores,
+                                       executor_memory=memory, **spark_args)
+        else:
+            raise ValueError("cluster_mode can only be local, yarn-client or standalone, "
+                            "but got: %s".format(cluster_mode))
```
Comment (hkvision) [inline python/orca/src/bigdl/orca/common.py:274]: need to modify this

File: python/orca/src/bigdl/orca/common.py [hunk 11]
```diff
@@ -183,79 +186,92 @@ def init_orca_context(cluster_mode="local", cores=2, memory="2g", num_nodes=1,
...
     spark_args = {}
     for key in ["conf", "spark_log_level", "redirect_spark_log"]:
         if key in kwargs:
             spark_args[key] = kwargs[key]
-    if cluster_mode == "spark-submit":
+    if cluster_mode is not None:
+        cluster_mode = cluster_mode.lower()
+    activate_sc = SparkContext._active_spark_context is not None
+    if activate_sc:
+        if cluster_mode is not None:
```
Comment (hkvision) [inline python/orca/src/bigdl/orca/common.py:206]: if cluster mode != "spark-submit" and cluster_mode is not None?

File: python/orca/src/bigdl/orca/common.py [hunk 12]
```diff
@@ -183,79 +186,92 @@ def init_orca_context(cluster_mode="local", cores=2, memory="2g", num_nodes=1,
...
+                                    executor_memory=memory, **spark_args)
+            else:
+                sc = init_spark_on_yarn(hadoop_conf=hadoop_conf,
+                                    conda_name=conda_env_name,
+                                    num_executors=num_nodes, executor_cores=cores,
+                                    executor_memory=memory, **spark_args)
+        elif cluster_mode.startswith("k8s"):  # k8s or k8s-client
+            if cluster_mode == "k8s-cluster":
+                raise ValueError('For k8s-cluster mode, '
+                                'please submit the application via spark-submit instead')
```
Comment (hkvision) [inline python/orca/src/bigdl/orca/common.py:252]: via spark-submit and use the default cluster_mode instead

**Comment by hkvision** [comment]
> > @jason-dai Shall we directly call init_nncontext in init_orca_context? As dllib has all these things now. It is better if we directly call init_nncontext and modify in init_nncontext to avoid duplicate work? cc @dding3 @qiuxin2012
> 
> Yes - but probably in a separate PR

Okay, so we merge this PR for the single purpose to change the default cluster mode to None first. Raise an issue to track the merge: https://github.com/intel-analytics/BigDL/issues/3414

File: python/orca/src/bigdl/orca/common.py [hunk 13]
```diff
@@ -158,22 +158,29 @@ def get_ray_context():
...
     """
     Creates or gets a SparkContext for different Spark cluster modes (and launch Ray services
     across the cluster if necessary).
 
     :param cluster_mode: The mode for the Spark cluster. One of "local", "yarn-client",
-           "yarn-cluster", "k8s-client", "standalone" and "spark-submit". Default to be "local".
+           "yarn-cluster", "k8s-client" and "standalone". Default to be None.
+           
+           For "yarn-client" and "yarn-cluster", you are supposed to use conda environment 
+           and set hadoop conf dir.
```
Comment (hkvision) [inline python/orca/src/bigdl/orca/common.py:171]: set the environment variable HADOOP_CONF_DIR

File: python/orca/src/bigdl/orca/common.py [hunk 14]
```diff
@@ -158,22 +158,29 @@ def get_ray_context():
...
-def init_orca_context(cluster_mode="local", cores=2, memory="2g", num_nodes=1,
+def init_orca_context(cluster_mode=None, cores=2, memory="2g", num_nodes=1,
                       init_ray_on_spark=False, **kwargs):
     """
     Creates or gets a SparkContext for different Spark cluster modes (and launch Ray services
     across the cluster if necessary).
 
     :param cluster_mode: The mode for the Spark cluster. One of "local", "yarn-client",
-           "yarn-cluster", "k8s-client", "standalone" and "spark-submit". Default to be "local".
+           "yarn-cluster", "k8s-client" and "standalone". Default to be None.
```
Comment (hkvision) [inline python/orca/src/bigdl/orca/common.py:168]: Default to be None and in this case there is supposed to be an existing SparkContext in your application.

File: python/orca/src/bigdl/orca/common.py [hunk 15]
```diff
@@ -183,79 +190,92 @@ def init_orca_context(cluster_mode="local", cores=2, memory="2g", num_nodes=1,
...
+        if cluster_mode == "local":
+            if num_nodes > 1: 
+                warnings.warn("For Spark local mode, num_nodes should be 1, but got " 
+                + repr(num_nodes) + ", ignored", Warning)
+            os.environ["SPARK_DRIVER_MEMORY"] = memory
+            if "python_location" in kwargs:
+                spark_args["python_location"] = kwargs["python_location"]
+            from bigdl.dllib.nncontext import init_spark_on_local
+            sc = init_spark_on_local(cores, **spark_args) 
+        elif cluster_mode.startswith("yarn"):  # yarn or yarn-client
```
Comment (hkvision) [inline python/orca/src/bigdl/orca/common.py:225]: # yarn, yarn-client or yarn-cluster

**Review by hkvision — APPROVED** [review]
